### PR TITLE
Dev/annagrin/remove explicit not null constructor

### LINKED ISSUE #699: 2fc94db made gsl::not_null constructor explicit, why?
Since 2fc94db3ebfb1b066edeafac1837f34d6111bff4, this program does not work anymore without explicit cast to not_null (here tested with clang 5.0.2 on linux with -std=c++14):
```
#include <gsl/pointers>
#include <iostream>

void f(gsl::not_null<const int *> i) {
  std::cout << *i << '\n';
}

int main() {
  int i = 42;
  f(&i); // no matching function for call to 'f'
         // note: candidate function not viable: no known conversion from 'int *' to 'gsl::not_null<const int *>' for 1st argument
  return 0;
}
```

If this is really wanted, could you explain why?

Kind regards

### PR DESCRIPTION
**Pull request by annagrin**
Removed explicit `not_null` constructor, `sloppy_not_null`, added `strict_not_null`

We added explicit `not_null` constructor in version 2.0.0.
It proved very difficult to switch to the new version for
large code bases that adopted previous versions of GSL,
due to `not_null` used extensively in the code. Still, using
explicit constructor is very beneficial for new code, since
it encourages better API design and make null checks intentional.

To resolve the issue, this change:
- removes explicit keyword from `not_null` constructor
- removes unneeded `sloppy_not_null `type
- adds `strict_not_null` type to behave the same way as v2 `not_null
`- updates tests

this closes #699

### CODE REVIEW

**Comment by annagrin** [comment]
@neilmacintosh could you please review?

File: tests/notnull_tests.cpp [hunk 1]
```diff
@@ -142,30 +142,73 @@ bool helper_const(not_null<const int*> p)
...
+        not_null<int*> x = &t;
+        helper(&t);
+        helper_const(&t);
+
+        CHECK(*x == 42);
+    }
+
+    {
+        int t = 42;
+        int* p = &t;
```
[the remark this answers is not in the document]
Comment (annagrin) [reply tests/notnull_tests.cpp]: yeah, you are right, the idea was that &i cover those tests, but having them added would be nice. Will change.

File: tests/test.cpp [hunk 2]
```diff
@@ -20,7 +20,7 @@
 
 // blanket turn off warnings from CppCoreCheck from catch
 // so people aren't annoyed by them when running the tool.
-#pragma warning(disable : 26401 26409 26415 26418 26426 26429 26432 26433 26434 26435 26436 26439 26440 26443 26444 26446 26447 26451 26460 26461 26466 26472 26481 26482 26485 26492 26493 26494 26495 26496 26497) // from catch
+#pragma warning(disable : 26401 26409 26415 26418 26426 26429 26432 26433 26434 26435 26436 26439 26440 26443 26444 26446 26447 26451 26455 26456 26460 26461 26466 26472 26481 26482 26485 26492 26493 26494 26495 26496 26497) // from catch
```
[the remark this answers is not in the document]
Comment (annagrin) [reply tests/test.cpp]: yes, i don't think we care about warnings in test.cpp itself... will change,

File: tests/strict_notnull_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+{
+    {
+        // raw ptr <-> strict_not_null
+        int x = 42;
+
+#ifdef CONFIRM_COMPILATION_ERRORS
+        strict_not_null<int*> snn = &x;
+        strict_helper(&x);
+        strict_helper_const(&x);
+        strict_helper(return_pointer());
```
[the remark this answers is not in the document]
Comment (annagrin) [reply tests/strict_notnull_tests.cpp]: sure, makes sence

File: include/gsl/pointers [hunk 4]
```diff
@@ -189,6 +189,100 @@ struct hash<gsl::not_null<T>>
...
+    {}
+
+    template <typename U, typename = std::enable_if_t<std::is_convertible<U, T>::value>>
+    constexpr strict_not_null(const strict_not_null<U>& other) :
+        not_null<T>(other)
+    {}
+
+    strict_not_null(strict_not_null&& other) = default;
+    strict_not_null(const strict_not_null& other) = default;
+    strict_not_null& operator=(const strict_not_null& other) = default;
```
[the remark this answers is not in the document]
Comment (annagrin) [reply include/gsl/pointers]: seems like an old omission here... will change as well, but maybe in a different PR.

File: include/gsl/pointers [hunk 5]
```diff
@@ -72,13 +72,13 @@ public:
     static_assert(std::is_assignable<T&, std::nullptr_t>::value, "T cannot be assigned nullptr.");
 
     template <typename U, typename = std::enable_if_t<std::is_convertible<U, T>::value>>
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply include/gsl/pointers]: `= delete` doesn't mean that the pertinent function doesn't participate in overload resolution, it means the program is ill-formed if the function is chosen by overload resolution.

File: include/gsl/pointers (hunk 4 above)
[the remark this answers is not in the document]
Comment (annagrin) [reply include/gsl/pointers]: "`strict_not_null& operator=(const strict_not_null&& other)` is a not special member function which can be defaulted" - so this seem to be an omission from the language itself:)

File: include/gsl/pointers (hunk 5 above)
[the remark this answers is not in the document]
Comment (annagrin) [reply include/gsl/pointers]: @CaseyCarter thanks, this exactly what I was searching for!

**Comment by annagrin** [comment]
thanks @neilmacintosh and @CaseyCarter for reviews!

**Pull request merged**